ogr/ogrsf_frmts/miramon/mm_gdal_functions.c

```diff
@@ -0,0 +1,2801 @@
+{
+    camp->reserved_2[MM_OFFSET_RESERVED2_EXTENDED_NAME_SIZE] =
+        (MM_BYTE)strlen(camp->FieldName);
+    return MM_DonaBytesNomEstesCamp(camp);
+}
+
+static MM_ACUMULATED_BYTES_TYPE_DBF
+MM_CalculateBytesExtendedFieldNames(const struct MM_DATA_BASE_XP *bd_xp)
+{
+    MM_ACUMULATED_BYTES_TYPE_DBF bytes_acumulats = 0;
+    MM_EXT_DBF_N_FIELDS i_camp;
+
+    for (i_camp = 0; i_camp < bd_xp->nFields; i_camp++)
+    {
+        if (MM_VALID_EXTENDED_DBF_NAME ==
+            MM_ISExtendedNameBD_XP(bd_xp->pField[i_camp].FieldName))
+            bytes_acumulats +=
+                MM_CalculateBytesExtendedFieldName(bd_xp->pField + i_camp);
+    }
+
+    return bytes_acumulats;
+}
+
+static MM_FIRST_RECORD_OFFSET_TYPE
+MM_CalculateBytesFirstRecordOffset(struct MM_DATA_BASE_XP *bd_xp)
+{
+    if (bd_xp)
+        return (32 + 32 * bd_xp->nFields + 1 +
+                MM_CalculateBytesExtendedFieldNames(bd_xp));
+    return 0;
+}
+
+static void MM_CheckDBFHeader(struct MM_DATA_BASE_XP *bd_xp)
+{
+    struct MM_FIELD *camp;
+    MM_EXT_DBF_N_FIELDS i;
+    MM_BOOLEAN cal_DBF_estesa = FALSE;
+
+    bd_xp->BytesPerRecord = 1;
+    for (i = 0, camp = bd_xp->pField; i < bd_xp->nFields; i++, camp++)
+    {
+        camp->AcumulatedBytes = bd_xp->BytesPerRecord;
+        bd_xp->BytesPerRecord += camp->BytesPerField;
+        if (camp->DesiredWidth == 0)
+            camp->DesiredWidth = camp->OriginalDesiredWidth =
+                MM_GetDefaultDesiredDBFFieldWidth(camp);  //camp->BytesPerField;
+        if (camp->FieldType == 'C' &&
+            camp->BytesPerField > MM_MAX_AMPLADA_CAMP_C_DBF_CLASSICA)
+            cal_DBF_estesa = TRUE;
+        if (MM_VALID_EXTENDED_DBF_NAME ==
+            MM_ISExtendedNameBD_XP(camp->FieldName))
+            cal_DBF_estesa = TRUE;
+    }
+
+    bd_xp->FirstRecordOffset = MM_CalculateBytesFirstRecordOffset(bd_xp);
+
+    if (cal_DBF_estesa || bd_xp->nFields > MM_MAX_N_CAMPS_DBF_CLASSICA ||
+        bd_xp->nRecords > UINT32_MAX)
+        bd_xp->dbf_version = (MM_BYTE)MM_MARCA_VERSIO_1_DBF_ESTESA;
+    else
+        bd_xp->dbf_version = MM_MARCA_DBASE4;
+}
+
+static void
+MM_InitializeOffsetExtendedFieldNameFields(struct MM_DATA_BASE_XP *bd_xp,
+                                           MM_EXT_DBF_N_FIELDS i_camp)
+{
+    memset((char *)(&bd_xp->pField[i_camp].reserved_2) +
+               MM_OFFSET_RESERVAT2_OFFSET_NOM_ESTES,
+           0, 4);
+}
+
+static void
+MM_InitializeBytesExtendedFieldNameFields(struct MM_DATA_BASE_XP *bd_xp,
+                                          MM_EXT_DBF_N_FIELDS i_camp)
+{
+    memset((char *)(&bd_xp->pField[i_camp].reserved_2) +
+               MM_OFFSET_RESERVED2_EXTENDED_NAME_SIZE,
+           0, 1);
+}
+
+static short int MM_return_common_valid_DBF_field_name_string(char *cadena)
+{
+    char *p;
+    short int error_retornat = 0;
+
+    if (!cadena)
+        return 0;
+    //strupr(cadena);
+    for (p = cadena; *p; p++)
+    {
+        (*p) = (char)toupper(*p);
```

Comment:
to avoid potential undefined behaviour with toupper:
```suggestion
        (*p) = (char)toupper((unsigned char)*p);
```